### ISSUE TITLE
Handle Generic method in dataclass_transform's field_specifiers

### ISSUE THREAD

**Issue by lukaspiatkowski** [microsoft/pyright#5882]
**Is your feature request related to a problem? Please describe.**

This is also related to https://github.com/python/mypy/issues/16031
I would like this code to be possible to write:
```
class F(Generic[T]):
    @staticmethod
    def field(factory: Callable[[T], V]) -> V: ...

class G(Generic[T]):
    @staticmethod
    @dataclass_transform(field_specifiers=(F[T].field, ))
    def foo(c): ...

class UserModel:
	id: str
	name: str

class ItemModel:
	id: str
	price: int

@G[UserModel].foo
class User:
	id: str = F[UserModel].field(factory = lamda user: user.id)
	name: str = F[UserModel].field(factory = lamda user: user.name)
	more_data: str

@G[ItemModel].foo
class User:
	id: str = F[ItemModel].field(factory = lamda item: item.id)
	price: int = F[ItemModel].field(factory = lamda item: item.price)
	more_data: str
```

And this to be flagged by the type checker:
```
@G[UserModel].foo
class User:
    # Error: F[ItemModel].field found where F[UserModel].field expected
	id: str = F[ItemModel].field(factory = lamda item: item.id)
	# Error: no field "price" found for user: UserModel
	name: str = F[UserModel].field(factory = lamda user: user.price)
```

The main idea here is to be able to force consistent `field` type for the decorated class without writing a new decorator for each use-case, because creating `@decorator_for_user` and `@decorator_for_item` in the above example would be too verbose.

**Describe the solution you’d like**

`@dataclass_transform` is honored as a decorator on Generic's method and one can pass a Generic method inside `field_specifiers`. Unfortunately the TypeVar on which the decorator is bound doesn't seem to bind the field specifier. See this example [playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAF329GQC_41RuYrDMBDt_RVDtrHA6yqVIUtCIPkB48YEIVvjjUCWjDRekr_f8ZGj2GJViac37xh1wfdA98G4bzD94APBUVmrGosZnNFhMG0G5X3ASoUMtCLVWhWjpKBc7Hzok6SE3YORbsqNSKp3oGIgSeYhOKWrZF1eRJEAn30kRabtka5ez4jGDjqDVqedasmHe_GMVPNcBtVFwOcXVIvAdPI8f3qc_-Ox_6NIOpvKOGBr-BriLj2xRj7DmRCvcN6nbTFtDecg0-WVJSCNwUHLgSZywzu4FRApzFzjaKGutO2WeftzzTg7eb-WOCykG3fnF17nqWaFNctjLzvWFvAB5dVEiFc_Wg0NgnKAIXj-rWWI845OZ9CMxMhk9EbGG9cl1KBHBPKwJOFMAX9QWTl1Sw-5lMYZklIkv7EDe7UxAgAA). So in the above code using `G[UserModel].foo` currently doesn't force usage of `F[UserModel].field` as field specifier, it is as if `F[Any].field` is allowed. My proposition is to make it so that the field specifier is bound by the outer Generic TypeVar.

**Comment by erictraut** [comment]
I give you points for creativity, but your use case is far outside of the intended uses for `dataclass_transform`. Sorry, but it's unlikely that we'd add support for this feature. I recommend that you look for other solutions to the problem you're trying to solve.

**Comment by lukaspiatkowski** [comment]
Thanks for you appreciation @erictraut! :) Just in case it matters, here is another example where it would be useful to have this feature:
```
class F(Generic[T]):
    @staticmethod
    def field(permission: T): ...

class G(Generic[T]):
    @staticmethod
    @dataclass_transform(field_specifiers=(F[T].field, ))
    def foo(c): ...

class UserPermissions(Enum):
    READ_PUBLIC = 0
    READ_PRIVATE = 1

class ItemPersmission(Enum):
    READ_OWNER = 0
    READ_PRICE = 0

@G[UserPermissions].foo
class User:
	name: str = F[UserPermissions].field(permission = UserPermission.READ_PUBLIC)
	age: int = F = F[UserPermissions].field(permission = UserPermission.READ_PRIVATE)

@G[ItemPersmission].foo
class Item:
    owner: User = F[ItemPersmission].field(permission = ItemPersmission.READ_OWNER)
	price: int = F[ItemPersmission].field(permission = ItemPersmission.READ_PRICE)
	more_data: str
```

**Comment by erictraut** [comment]
Like I said, it's very unlikely we'd ever implement support for this. The field specifiers are arguments passed to the decorator. They are evaluated at the time the decorator is applied. The function or class that the `dataclass_transform` is applied to is not generic in your example, so it cannot be later specialized once the value of `T` is supplied for the containing class.

Here's an alternative approach (based on your most recent code sample) that works with pyright and is within the bounds of the intended use cases for `dataclass_transform`. It exposes a bug in mypy, but that should be easy to address.

```python
from typing import Any, Generic, TypeVar, dataclass_transform
from enum import Enum

T = TypeVar("T")

class PermissionField(Generic[T], Any):
    def __init__(self: Any, permission: T) -> None:
        ...

class UserPermission(Enum):
    READ_PUBLIC = 0
    READ_PRIVATE = 1

@dataclass_transform(field_specifiers=(PermissionField[UserPermission],))
def user_permission_dataclass(c):
    ...

@user_permission_dataclass
class User:
    name: str = PermissionField(permission=UserPermission.READ_PUBLIC)
    age: int = PermissionField(permission=UserPermission.READ_PRIVATE)

class ItemPersmission(Enum):
    READ_OWNER = 0
    READ_PRICE = 1

@dataclass_transform(field_specifiers=(PermissionField[ItemPersmission],))
def item_permission_dataclass(c):
    ...

@item_permission_dataclass
class Item:
    owner: User = PermissionField(permission=ItemPersmission.READ_OWNER)
    price: int = PermissionField(permission=ItemPersmission.READ_PRICE)
    more_data: str
```
